lgc/patch/PatchReadfirstlaneOpt.cpp

```diff
@@ -0,0 +1,204 @@
+bool PatchReadfirstlaneOpt::runOnFunction(Function &function) {
+  LLVM_DEBUG(dbgs() << "Run the pass Patch-Readfirstlane-Opt\n");
+
+  m_divergenceAnalysis = &getAnalysis<LegacyDivergenceAnalysis>();
+  visit(function);
+
+  const bool changed = pushUpReadfirstlane();
+
+  return changed;
+}
+
+// =====================================================================================================================
+// Specify what analysis passes this pass depends on.
+//
+// @param [in,out] analysisUsage : The place to record our analysis pass usage requirements.
+void PatchReadfirstlaneOpt::getAnalysisUsage(AnalysisUsage &analysisUsage) const {
+  analysisUsage.addRequired<LegacyDivergenceAnalysis>();
+}
+
+// =====================================================================================================================
+// Visits "call" instruction.
+//
+// @param callInst : The instruction
+void PatchReadfirstlaneOpt::visitCallInst(CallInst &callInst) {
+  // Collect the intial uniform operands as the start point for pushing the corresponding readfirstlanes higher
+  if (auto intrinsicCall = dyn_cast<IntrinsicInst>(&callInst)) {
+    if (intrinsicCall->getIntrinsicID() == Intrinsic::amdgcn_readfirstlane) {
+      if (auto extract = dyn_cast<ExtractElementInst>(callInst.getArgOperand(0)))
+        m_intialUniforms.insert(extract->getOperand(0));
+    }
+  }
+}
+
+// =====================================================================================================================
+// Push the collected initial readfirstlanes as high as they can to transform more VGPR instructions into SGPR
+// instructions
+bool PatchReadfirstlaneOpt::pushUpReadfirstlane() {
+  if (m_intialUniforms.empty())
+    return false;
+
+  // Represent the operand info for checking if it is can be uniform
+  struct OperandInfo {
+    // The operand to be checked
+    Value *operand;
+    // Whether the operand is from m_intialUniforms
+    bool isInitialUniform;
+  };
+  // Collect the operands info by tracking back
+  std::stack<OperandInfo> operandInfoStack;
+
+  // Set initial tested operands based on m_initialUniforms
+  for (auto initialUniform : m_intialUniforms) {
+    OperandInfo operandInfo = {initialUniform, true};
+    if (auto load = dyn_cast<LoadInst>(initialUniform)) {
+      // The operand is from descriptor index
+      if (auto bitcast = dyn_cast<BitCastInst>(load->getOperand(0))) {
+        if (auto gep = dyn_cast<GetElementPtrInst>(bitcast->getOperand(0))) {
+          if (!gep->hasAllConstantIndices() && gep->getNumIndices() == 1) {
+            operandInfo.operand = *gep->idx_begin();
+            operandInfoStack.push(operandInfo);
+          }
+        }
+      }
+    } else
+      operandInfoStack.push(operandInfo);
+  }
+
+  if (operandInfoStack.empty())
+    return false;
+
+  // Collect the visited operands to keep the operand checked only once
+  llvm::DenseSet<Value *> visitedOpernads;
```

Comment:
Spelling: visitedOperands

Also, `llvm::` is not necessary here due to `using namespace` above.

Reply:
Sorry for the typo. Fixed and removed llvm::